core/trino-main/src/test/java/io/trino/operator/scalar/TestJsonStringArrayExtractScalar.java

```diff
@@ -0,0 +1,145 @@
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+import static org.junit.jupiter.api.parallel.ExecutionMode.CONCURRENT;
+
+@TestInstance(PER_CLASS)
+@Execution(CONCURRENT)
+class TestJsonStringArrayExtractScalar
+{
+    private QueryAssertions assertions;
+
+    @BeforeAll
+    public void init()
+    {
+        assertions = new QueryAssertions();
+    }
+
+    @AfterAll
+    public void teardown()
+    {
+        assertions.close();
+        assertions = null;
+    }
+
+    @Test
+    void testEmptySliceExtract()
+    {
+        assertTrinoExceptionThrownBy(() -> assertions.expression("transform(cast(json_parse(a) as array<json>), x -> json_extract_scalar(\"x\", '$.name'))")
+                .binding("a", "''").evaluate())
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessageContaining("Cannot cast to array(varchar). Expected a json array, but got null");
+    }
+
+    @Test
+    void testEmptyStringExtract()
+    {
+        failedExtract("", "$");
+    }
+
+    @Test
+    void testObjectStringExtract()
+    {
+        failedExtract("{}", "$");
+    }
+
+    @Test
+    void testScalarStringExtract()
+    {
+        failedExtract("1", "$");
+    }
+
+    @Test
+    void testEmptyArrayExtract()
+    {
+        Result result = extract("[]", "$.name");
+        assertThat(result.value()).isEqualTo(List.of());
+    }
+
+    @Test
+    void testScalarArrayExtract()
+    {
+        Result result = extract("[1, \"abc\", 2.2]", "$.name");
+        assertThat(result.value()).isEqualTo(asList(null, null, null));
+        result = extract("[1, \"abc\", 2.2]", "$");
+        assertThat(result.value()).isEqualTo(asList("1", "abc", "2.2"));
+    }
+
+    @Test
+    void testEmptyElementsArrayExtract()
+    {
+        Result result = extract("[null, [], {}]", "$.name");
+        assertThat(result.value()).isEqualTo(asList(null, null, null));
+    }
+
+    @Test
+    void testExtract()
+    {
+        Result result = extract("[{\"name\": \"a\"}, {\"name\": \"b\"}, [], {}]", "$.name");
+        assertThat(result.value()).isEqualTo(asList("a", "b", null, null));
+    }
+
+    @Test
+    void testMixedExtract()
+    {
+        Result result = extract("[{\"name\": \"a\"}, 1]", "$.name");
+        assertThat(result.value()).isEqualTo(asList("a", null));
+    }
+
+    private Result extract(String json, String jsonPath)
+    {
+        Result transformationResult = extractTransformationResult(json, jsonPath);
+        Result jsonStringResult = extractJsonStringArrayExtractResult(json, jsonPath);
+        assertThat(transformationResult.value()).isEqualTo(jsonStringResult.value());
+        return jsonStringResult;
+    }
+
+    private Result extractTransformationResult(String json, String jsonPath)
+    {
+        // inlined value for json_parse does not trigger optimization for cast + json_parse
```

Comment:
It is inlined as array<json> which does not trigger the rule.

Reply:
I'm not a great fan of these tests. The problem is that we don't know if `extractTransformationResult` actually used optimized expression nor that `extractJsonStringArrayExtractResult` didn't use it.

Some kind of test similar to `TestLogicalPlanner` would be useful.

Now if you look at `io.trino.sql.query.QueryAssertions.ExpressionAssertProvider#run` you can see that it uses `execute`. However, you could use `executeWithPlan` to get plan also. Then such plan could be used to extract `IrExpressions` and return them in `ExpressionAssertProvider.Result`. Then you could check that your optimization actually executed (or didn't)

Reply:
I was thinking of such assurance if optimization works.